### ISSUE TITLE
Split the wasm file

### ISSUE THREAD

**Issue by NateLing** [bevyengine/bevy#4555]
Bevy output wasm is big , [3d-scene](https://bevyengine.org/examples/3d/3d-scene/) wasm file was 14MB, gzip size is 3.6.

Is it possible to split a single wasm file into different modules to better utilize browser network concurrency?

**Comment by jakobhellermann** [comment]
I don't think this is possible in wasm. There's an open issue with some discussion <https://github.com/WebAssembly/design/issues/1166>, but I can't find anything else. On other platforms you can also split the debuginfo into a separate file, but that doesn't seem possible on wasm. On a release build you can just remove the debug info instead which further cuts file size: <https://github.com/bevyengine/bevy/issues/3978#issuecomment-1044333040>. With all the options in that comment I can go from `17mb` to `1.7mb` which might be enough for you.

**Comment by NateLing** [comment]
> I don't think this is possible in wasm. There's an open issue with some discussion <https://github.com/WebAssembly/design/issues/1166>, but I can't find anything else. On other platforms you can also split the debuginfo into a separate file, but that doesn't seem possible on wasm. On a release build you can just remove the debug info instead which further cuts file size: <https://github.com/bevyengine/bevy/issues/3978#issuecomment-1044333040>. With all the options in that comment I can go from `17mb` to `1.7mb` which might be enough for you.

Yes,But there's an other way to resolve this. We can split some standalone module, and then use  webworker  to post message.

**Comment by druskus20** [comment]
Wasm code splitting is, to the best of my knowledge, not implemented in Rust. 
There's a 2018 issue discussing it https://github.com/rustwasm/team/issues/52

And it is implemented for C++ https://github.com/WebAssembly/binaryen/blob/102c3633d2378457dae1f5e239fd63ad80eefb92/src/tools/wasm-split/wasm-split.cpp#L339

In the case of web at least, the limitation is still lazy loading the modules. https://emscripten.org/docs/optimizing/Module-Splitting.html#running-on-the-web

**Comment by jbms** [comment]
See this issue where I posted a prototype that addresses exactly this: https://github.com/rustwasm/wasm-bindgen/issues/3939

**Comment by daxpedda** [comment]
To at least get rid of the debug info, I'm using [`wasm2map`](https://github.com/mtolmacs/wasm2map) to generate a source map and then strip the debug info from the Wasm file.
